### DISCUSSION TITLE
[2주차]_4장_처리율 제한 장치의 설계

### DISCUSSION THREAD

**Discussion by minju412** [11th-SSAFY-19/large-scale-system-design#11] [General]
.

**Comment by chwangmin** [comment]
> 깜짝 세일같은 이벤트 때문에 트래픽이 급증할 때 처리율 제한 장치가 비효율적으로 동작

여기서 순간 트래픽 집중이 왜 처리율 제한 장치가 비효율적으로 동작하는 이유가 궁금합니다.
큐에 요청이 초과되는 오버플로우 때문에, 처리율 제한 장치가 비효율적이라고 판단하면 되는 걸까요?
그리고 그 해결방법이 왜 토큰버킷 알고리즘 일까요?

**Reply by minju412** [reply]
트래픽 급증 시 비효율적인 이유: 
전통적인 처리율 제한 장치는 초당 요청 수를 고정된 값으로 제한합니다. 이 방식은 일정한 트래픽에는 유효하지만, 갑작스러운 트래픽 급증 상황에서는 제한을 걸어버리기 때문에 사용자가 대기하거나 요청이 거부되는 경우가 발생할 수 있습니다.

토큰버킷 알고리즘이 해결 방법인 이유: 
토큰 버킷 알고리즘은 버킷에 일정량의 토큰을 저장하고, 요청이 들어올 때마다 토큰을 소모하는 방식입니다. 버킷에 미리 일정량의 토큰이 쌓여 있을 수 있기 때문에, 갑작스러운 트래픽 급증에 대해 유연하게 대응할 수 있습니다. 즉, 한 번에 많은 요청이 들어와도 버킷에 토큰이 충분하면 요청을 처리할 수 있습니다.

**Reply by minju412** [reply]
+ 동작이 가장 간단해서

**Reply by Hyunuk17** [reply]
GPT

토큰 버킷 알고리즘(Token Bucket Algorithm)은 대규모 트래픽 상황에서 다음과 같은 이유로 유리합니다:

유연한 대역폭 제어:
- 토큰 버킷 알고리즘은 설정된 비율로 토큰을 추가하여 평상시의 트래픽 양을 제어하면서도 일정 기간 동안 버킷에 쌓인 토큰을 통해 일시적인 트래픽 폭증을 허용합니다. 이로 인해 일관된 대역폭 사용을 보장하면서도 급격한 트래픽 변동을 유연하게 처리할 수 있습니다.

버스트 트래픽 허용:
- 버스트 트래픽은 짧은 시간 동안 집중적으로 발생하는 트래픽을 의미합니다. 토큰 버킷 알고리즘은 버킷에 토큰이 충분히 쌓여 있을 경우, 이러한 버스트 트래픽을 일시적으로 허용하여 서비스 품질을 높이고 사용자 경험을 향상시킵니다.

간단한 구현과 낮은 오버헤드:
- 토큰 버킷 알고리즘은 구현이 비교적 간단하고, CPU나 메모리와 같은 시스템 자원을 많이 소모하지 않습니다. 대규모 트래픽 상황에서도 시스템 성능에 큰 영향을 미치지 않으면서 효과적으로 트래픽을 제어할 수 있습니다.

공정한 서비스 제공:
- 토큰 버킷 알고리즘은 일정한 비율로 토큰을 추가함으로써, 시간이 지남에 따라 각 사용자나 세션이 공정하게 네트워크 자원을 사용할 수 있도록 합니다. 이를 통해 특정 사용자나 세션이 과도하게 자원을 독점하는 상황을 방지할 수 있습니다.

QoS(Quality of Service) 보장:
- 네트워크에서 다양한 서비스의 품질을 보장하기 위해 토큰 버킷 알고리즘을 활용할 수 있습니다. 예를 들어, 우선 순위가 높은 트래픽은 더 많은 토큰을 할당받아 지연 없이 전달되도록 하고, 우선 순위가 낮은 트래픽은 여유가 있을 때만 전달되도록 설정할 수 있습니다.

통제 가능한 트래픽 속도:
- 네트워크 관리자는 토큰 추가 속도와 버킷 크기를 조절하여 트래픽 속도를 세밀하게 제어할 수 있습니다. 이를 통해 네트워크의 전반적인 성능을 최적화하고 과부하를 방지할 수 있습니다.

이러한 이유들로 인해 토큰 버킷 알고리즘은 대규모 트래픽 상황에서 안정적이고 효율적인 트래픽 관리 솔루션으로 널리 사용됩니다.

**Comment by hoshogi** [comment]
<img width="409" alt="스크린샷 2024-07-04 오후 8 14 43" src="https://github.com/11th-SSAFY-19/large-scale-system-design/assets/71253970/6bf89766-b203-4d71-8610-bf32a9da2add">


이동 윈도 로깅 알고리즘 동작 과정 그림 중 4번에서 만약 1:01:41에 추가 요청이 온다면 이 요청은 거부되는 건지 궁금합니다.
4번 버킷에 이미 로그 2개가 있어서 거부될 수 있을 것 같아서요!

**Reply by minju412** [reply]
현재 시점(1:01:41)에서 1분 이내 요청이 이미 두 개 이기 때문에 추가 요청은 거부됩니다!

**Reply by chwangmin** [reply]
거부하는 관점에서 보면 이해하기 좋을 거 같아요

**Comment by minju412** [comment]
<img width="1211" alt="스크린샷 2024-07-04 오후 9 02 43" src="https://github.com/11th-SSAFY-19/large-scale-system-design/assets/59405576/949c1cf9-35ef-4506-8c8f-762293d86bf1">
캐시와 레디스를 따로 분리해야하는 이유가 있을까요? 처리율 제한 규칙을 redis에 넣으면 안될까요?

**Reply by minju412** [reply]
여러개의 미들웨어가 있을 때 미들웨어 안에 있는 캐시를 표현하는 것 같습니다. -광민-

**Reply by minju412** [reply]
목적의 분리, 성능 개선 두 가지 이유가 있습니다.
- 목적의 분리
  - 캐시는 자주 조회되지만 변경이 빈번하지 않은 데이터 저장에 사용
  - 처리율 제한 규칙은 매우 빈번하게 업데이트 됨
- 성능 개선
  - 처리율 제한은 각 요청마다 레디스에 접근해서 데이터를 업데이트 해야 함 -> 레디스 서버에 부하
  - 캐시와 처리율 제한 규칙을 분리하면, 캐시로 인한 레디스 부하를 줄일 수 있다.

**Comment by minju412** [comment]
토큰 버킷 알고리즘에서 api마다 버킷을 생성해야 하는데, api가 많으면 성능 문제가 발생하지 않을까요? -혁진-

**Reply by minju412** [reply]
키 값이 너무 많은 경우 토큰 버킷 알고리즘을 사용하는 것은 부적절한 것 같다. -현재-

**Reply by minju412** [reply]
redis를 사용하면 key-value로 접근하기 때문에 빠를 것 같다. -호석-

**Reply by gurwls0122** [reply]
사용자가 많고, API가 여러개 인경우
1. 샤딩 : API를 여러 그룹으로 나누고 각 그룹에 대해 하나의 버킷을 사용하여 API에 대한 처리율 제한을 그룹 단위로 관리할 수 있다.
2. 분산형 버킷 관리 : 중앙 집중식 버킷 관리 대신 분산형으로 관리하여 개별 서버나 노드가 각자 버킷을 유지하도록 할 수 있다.
등의 방법도 있다고 하네요..

키 값이 너무 많은 경우 토큰 버킷 알고리즘을 사용하는 것은 부적절하다 -> X. 여전히 성능 및 다양한 면에서 사용 가능하다. 하지만, 모든 API를 버킷으로 만드는 것이 아닌 버킷의 개수를 조정하는 유동적인 방법을 사용하는 것이 좋다.

**Comment by minju412** [comment]
Iptable 스크립트를 적용하는 곳은 어디인가요? -호석-
- OS단?
- 라우터?
- 게이트웨이?
- 로드밸런서?
- EC2?

**Reply by minju412** [reply]
시스템 부팅 시 자동으로 iptables 규칙을 로드하기 때문에 운영체제 단에서 적용됩니다.

**Comment by minju412** [comment]
누출버킷 알고리즘에서 버킷 없이 큐만 사용하면 안되나요? -광민-

**Reply by minju412** [reply]
버킷이 없으면 말이 안되는 것 같다. 버킷에 들어오는 요청을 비동기적으로 처리하기 위한 수단으로 큐를 사용하는 것

**Reply by minju412** [reply]
https://etloveguitar.tistory.com/127

**Reply by gurwls0122** [reply]
https://yozm.wishket.com/magazine/detail/1900/
저는 버킷자체가 큐로 구현되어 있다고 이해했는데 다른분들은 어떠신가요?

**Reply by minju412** [reply]
누출 버킷 알고리즘에서 버킷은 `패킷 처리기`로, 큐에 있는 패킷을 일정한 간격으로 처리하는 로직이라고 합니다. 
![image](https://github.com/user-attachments/assets/2e3c9afb-8164-435c-ac7a-f16ce7578618)

**Comment by hoshogi** [comment]
토큰 버킷 알고리즘과 고정 윈도우 카운터 알고리즘이 비슷한 것 같아서 차이점이 궁금합니다.

### 토큰 버킷 vs 고정 윈도우 카운터
- 버킷의 크기 == 하나의 윈도우에서 처리하는 요청의 수
- 토큰 공급률 == 윈도우 사이즈(타임라인)

**Reply by minju412** [reply]
둘은 버스트 트래픽 처리, 트래픽 제어 방식에 차이점이 있습니다.
- 버스트 트래픽 처리
  - 토큰 버킷 : 일정한 속도로 토큰을 추가하기 때문에 버스트 트래픽을 더 잘 처리
  - 고정 윈도우 : 일정한 시간 간격 동안의 요청 수 제한
- 트래픽 제어 방식
  - 토큰 버킷 : 일정한 속도로 토큰을 추가해 트래픽 제어
  - 고정 윈도우 : 일정한 시간 간격 동안의 요청 수를 제한

**Reply by chwangmin** [reply]
토큰 버킷은 소비, 고정 윈도우는 추가
토큰 버킷은 최대량을 크게 추가해서 순간 트래픽을 방지.
같은 속도로 토큰을 추가

**Reply by galug** [reply]
**토큰 버킷을 구현할 때 max-size를 아예 높여 놓고 주는 공급량을 max-size 보다 적게 설정할 시 좀 더 유의미한 구현이 되지 않을까 싶습니다. **

**Comment by gurwls0122** [comment]
이동 윈도 카운터 방식의 장점에서 메모리 효율이 좋다고 나와있는데,
직전 1분간 요청 수를 기억해야 하기 때문에, 이를 어떻게 관리하기에 따라서 기존 이동 윈도 로그 방식에서 타임스탬프를 기억하는 것과 비슷해질 수도 있다고 생각하는데 어떻게 관리하는지 궁금해요!

**Reply by minju412** [reply]
이동 윈도 카운터 방식은 두 개의 카운터만 사용하기 때문에 타임스탬프를 저장하는 것보다 메모리를 효율적으로 사용합니다.
